### PR TITLE
Add support for optional certificate validation

### PR DESCRIPTION
**Pull request by gautierdelorme**
This is required to support optional mTLS so that apps can use different authentication schemes (e.g. mTLS, cookies, tokens, etc...).
I initially named the new option `OnlyVerifyClientCertIfGiven` but this was confusing when `SkipClientCertValidation` is also enabled. I decided to go with `OnlyRequestClientCert` (and not `RequireClientCert` for example) since all the other option are opt-in by setting them to true so I would like to avoid changing this assumption with this new option. The other problem was at the Go API level `DownstreamValidation.RequireClientCert` default value would be `false` since that's the null value for booleans which would change the current behavior and I wanted to avoid implementing it as a pointer.

Signed-off-by: Gautier Delorme <gautier.delorme@gmail.com>

### CODE REVIEW

**Comment by gautierdelorme** [comment]
Do you think this change could make it to the 1.23 release?

**Review by tsaarni — COMMENTED** [review]
Looks very good to me! :+1:  Small comment about the field name

File: apis/projectcontour/v1/httpproxy.go
Comment (tsaarni) [reply apis/projectcontour/v1/httpproxy.go:1135]: My thinking was that I'd highlight the security implications first: after enabling this option the proxy will not enforce the client to present certificate to be validated, which I think `OptionalClientCertificate` could communicate bit clearer. You are right of course, that the point for the use case is that certificate is still requested from client. I'd keep that clearly stated in the description.

Comment (gautierdelorme) [reply apis/projectcontour/v1/httpproxy.go]: @tsaarni I'm sold 👍 Renamed in https://github.com/projectcontour/contour/pull/4796/commits/0f03cb1b58645144b6fe63b8e3ba149de9eeac32.

File: changelogs/unreleased/4796-gautierdelorme-minor.md
Comment (skriss) [inline changelogs/unreleased/4796-gautierdelorme-minor.md:3]:
```suggestion
By default, when client certificate validation is configured, client certificates are required.
However, some applications might support different authentication schemes.
```

**Review by skriss — CHANGES_REQUESTED** [review]
@gautierdelorme nice work, everything looks great to me, just one minor wording suggestion on the changelog. Thanks for the PR!

File: changelogs/unreleased/4796-gautierdelorme-minor.md
Comment (gautierdelorme) [reply changelogs/unreleased/4796-gautierdelorme-minor.md]: @skriss 👍 fixed in https://github.com/projectcontour/contour/pull/4796/commits/d910de6ea00b8f2d3e3e406736b82dca7894e926

Comment (skriss) [reply changelogs/unreleased/4796-gautierdelorme-minor.md]: Looks like "However, some applications might support different authentication schemes." got dropped, was that intended? I'd keep that sentence in for clarity.

Comment (gautierdelorme) [reply changelogs/unreleased/4796-gautierdelorme-minor.md]: Good catch, it wasn't intended. Fixed in https://github.com/projectcontour/contour/pull/4796/commits/45a2250fc53d087bb4b053479492e88df0fa9a6c.

**Review by skriss — APPROVED** [review]
LGTM pending CI, will leave for a second maintainer to approve though. Thanks again @gautierdelorme!

**Review by tsaarni — APPROVED** [review]
Great to have this feature, thank you for contributing this @gautierdelorme!

**Pull request merged**
